### ISSUE TITLE
Segfault on player thread shutdown

### ISSUE THREAD

**Issue by voyagerok** [mikebrady/shairport-sync#870]
Hi. I have been testing shairport-sync on arm device, and there is one particular scenario, when it crashes every time with segfault.

#### Steps to reproduce:
Connect to shairport-sync from Apple Music, play something, then connect to different audio sink (headphones, for example) without stopping/pausing playback.

#### shairport-sync -V
`3.3.1-OpenSSL-tinysvcmdns-stdout-pipe-metadata-sysconfdir:/usr/etc`

#### Config
<details><summary>config contents</summary>
<p>

```
// General Settings
general =
{
    name = "Shairport Sync %v on ast250";
    interpolation = "basic";
    output_backend = "pipe";
    mdns_backend = "tinysvcmdns";
    port = 5000; // Listen for service requests on this port
    interface = "wlan0";
}

// Parameters for the "pipe" audio back end, a back end that directs raw CD-style audio output to a pipe. No interpolation is done.
// For this section to be operative, Shairport Sync must have been built with the following configuration flag:
// --with-pipe
pipe =
{
        name = "/tmp/sharport-sync-audio"; // there is no default pipe name for the output
};

diagnostics =
{
        disable_resend_requests = "no"; // set this to yes to stop Shairport Sync from requesting the retransmission of missing packets. Default is "no".
        statistics = "no"; // set to "yes" to print statistics in the log
        log_verbosity = 0; // "0" means no debug verbosity, "3" is most verbose.
        log_show_time_since_startup = "no"; // set this to yes if you want the time since startup in the debug message -- seconds down to nanoseconds
        log_show_time_since_last_message = "yes"; // set this to yes if you want the time since the last debug message in the debug message -- seconds down to nanoseconds
        drop_this_fraction_of_audio_packets = 0.0; // use this to simulate a noisy network where this fraction of UDP packets are lost in transmission. E.g. a value of 0.001 would mean an average of 0.1% of pack
ets are lost, which is actually quite a high figure.
};

// How to deal with metadata, including artwork
// For this section to be operative, Shairport Sync must be built with at one (or more) of the following configuration flags:
// --with-metadata, --with-dbus-interface, --with-mpris-interface or --with-mqtt-client.
// In those cases, "enabled" and "include_cover_art" will both be "yes" by default
metadata =
{
        enabled = "no"; // set this to yes to get Shairport Sync to solicit metadata from the source and to pass it on via a pipe
        include_cover_art = "no"; // set to "yes" to get Shairport Sync to solicit cover art from the source and pass it via the pipe. You must also set "enabled" to "yes".
        pipe_name = "/tmp/shairport-sync-metadata";
        pipe_timeout = 5000; // wait for this number of milliseconds for a blocked pipe to unblock before giving up
};

```

</p>
</details>

### gdb output
[shairport-gdb.txt](https://github.com/mikebrady/shairport-sync/files/3314419/shairport-gdb.txt)

### shairport logs
[shairport-crash-log.txt](https://github.com/mikebrady/shairport-sync/files/3314424/shairport-crash-log.txt)

As I understood from logs and source code, program received TEARDOWN command from AppleMusic and started procedure of shutting down current player thread in function `player_stop()` in player.c.
```
  if (conn->player_thread) {
    debug(3, "player_thread cancel...");
    pthread_cancel(*conn->player_thread);
    debug(3, "player_thread join...");
    if (pthread_join(*conn->player_thread, NULL) == -1) {
      char errorstring[1024];
      strerror_r(errno, (char *)errorstring, sizeof(errorstring));
      debug(1, "Connection %d: error %d joining player thread: \"%s\".", conn->connection_number,
            errno, (char *)errorstring);
    } else {
      debug(3, "player_thread joined.");
    }
```
As one can see, there is call to `pthread_cancel()` among others, which can trigger several cleanup handlers. Perhaps they are cleaning up something they should not?

Another detail: crash does not happen (well, I could not reproduce it) on x86 system with Ubuntu on it. Only on arm device I mentioned earlier.

**Comment by voyagerok** [comment]
Perhaps it is also connected with ThreadSanitizer warnings from #869. I'll try to investigate more, but it will take time as I unfamiliar with source code.

**Comment by mikebrady** [comment]
Thanks for the report. I’ll see if I can reproduce it. It would be great to get the ThreadSanitizer working on it alright.

**Comment by mikebrady** [comment]
Trying to reproduce this behavior. I’ve set up Shairport Sync as indicated and using `aplay` to play music from the pipe. Using iTunes as a source. Switching output repeatedly between Shairport Sync and the Computer Speakers is working normally. Similarly using the Music app on iOS and switching repeatedly between iPhone and Shairport Sync without stopping is working fine. Am I doing something different to your setup?

**Comment by voyagerok** [comment]
Hi. I did some debugging and found out that disabling -fomit-frame-pointer optimization seems to help. I'm not ready to say what is a real source of a problem. Perhaps there is some kind of bug in toolchain.

The problem occurred during stack unwinding when cancellation of thread is requested. The arg passed to cancellation cleanup callback through the stack and somehow it got overwritten by the time callback was called. I considered some kind of stack corruption but did not succeed tracking it down.

**Comment by mikebrady** [comment]
Many thanks. I’ll keep trying.

**Comment by mikebrady** [comment]
Not getting any further, TBH. Running the test `development` version, 3.3.2d3, on a Raspberry Pi 3B Plus running the most recent Raspbian Lite:
```
Linux raspberryPi3BP 4.19.50-v7+ #896 SMP Thu Jun 20 16:11:44 BST 2019 armv7l GNU/Linux
```

Can't get it too misbehave. I'm missing a secret sauce...

**Comment by mikebrady** [comment]
I guess there is a slight chance that the bug was fixed in the update, which was to fix a double-free fault associated with generating the service name...

**Comment by voyagerok** [comment]
Well, the problem still occurres sometimes, even with supplied patch. I am more inclined to think that there is a bug in toolchain, and probably has nothing to do with shairport-sync. I will report more info tomorrow.

**Comment by voyagerok** [comment]
When compiled with optimization level `-O1` bug seems to go away. I'm going to use that as workaround and not to go deeper investigating this issue. So this issue can be closed, I think.

**Comment by voyagerok** [comment]
This is what I found out so far: 
`player_thread_cleanup_handler` got its argument overwritten sometimes from random places of code,, usually from `rtsp_conversation_thread_func` according to gdb - I used watchpoint on memory location, where argument was stored. 

Code looks valid, and honestly I have no idea what is wrong. Moreover, I tried to run program under memcheck tool and stiil have no clue  where is possible memory corruption source.

There is a possiblity that gcc generates wrong code  but I have no intentions to investigate that.

**Comment by mikebrady** [comment]
Many thanks for your investigations. I’ll keep digging...

**Comment by voyagerok** [comment]
@mikebrady of course I should mention that toolchain for my platfrom includes gcc 4.4.1 and glibc 2.11.1. They are quite outdated and may contain some bugs that have been fixed a long time ago.

**Comment by mikebrady** [comment]
Closing this issue, as suggested – thanks for the contribution. Please open a new issue if necessary.
